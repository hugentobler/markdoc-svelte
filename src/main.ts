import { basename, extname } from "path";

import Markdoc from "@markdoc/markdoc";
import type { Config, ParserArgs } from "@markdoc/markdoc";
import type { PreprocessorGroup } from "svelte/compiler";
import YAML from "yaml";

import { getComponentImports, extractUsedSvelteComponents } from "./components.ts";
import { handleValidationErrors } from "./errors.ts";
import { findFirstDirectory, makePathProjectRelative } from "./files.ts";
import log from "./logs.ts";
import loadPartials from "./partials.ts";
import render from "./render.ts";
import loadSchemas from "./schema.ts";
import type { Options } from "./types";

const validOptionKeys: (keyof Options)[] = [
  "extensions",
  "schema",
  "nodes",
  "tags",
  "variables",
  "functions",
  "partials",
  "components",
  "layout",
  "comments",
  "linkify",
  "typographer",
  "validationLevel",
];

/**
 * Creates a Svelte preprocessor for Markdoc files
 * @param {Options} options - Configuration options for the Markdoc preprocessor
 * @returns {PreprocessorGroup} A Svelte preprocessor for Markdoc files
 */
export const markdocPreprocess = (options: Options = {}): PreprocessorGroup => {
  // Warn about invalid options
  for (const key in options) {
    if (!validOptionKeys.includes(key as keyof Options)) {
      log.warn(
        `Invalid option "${key}" provided and ignored.  Please check the documentation for valid options.`,
      );
    }
  }

  // --- Set defaults ---
  const extensions =
    options.extensions && options.extensions.length > 0
      ? options.extensions
      : [".mdoc", ".md"];
  const schemaPaths = options.schema
    ? [makePathProjectRelative(options.schema)]
    : ["./markdoc", "./src/markdoc"];
  const nodes = options.nodes;
  const tags = options.tags;
  const variables = options.variables;
  const functions = options.functions;
  const partialsPath = options.partials
    ? makePathProjectRelative(options.partials)
    : undefined;
  const componentsPath = options.components || "$lib/components";
  const layoutPath = options.layout;
  const allowComments = options.comments ?? true;
  const linkify = options.linkify ?? false;
  const typographer = options.typographer ?? false;
  const validationLevel = options.validationLevel || "error";

  return {
    name: "markdoc-svelte",
    markup: async ({ content, filename }) => {
      // --- Check if file is a Markdoc file ---
      if (
        !filename ||
        !extensions.find((extension) => filename.endsWith(extension))
      )
        return;

      // --- Tokenization ---
      // Markdown-It options https://github.com/markdown-it/markdown-it?tab=readme-ov-file#init-with-presets-and-options
      const markdownItConfig = {
        linkify,
        typographer,
      };
      const tokenizer = new Markdoc.Tokenizer({
        allowIndentation: true,
        allowComments: allowComments,
        ...markdownItConfig,
      });
      const tokens = tokenizer.tokenize(content);

      // --- Parse to AST ---
      const parserConfig: ParserArgs = {
        file: filename, // Debugging
        location: true, // Debugging
        slots: false, // TODO: Add support for slots?
      };
      const ast = Markdoc.parse(tokens, parserConfig);

      // --- Parse Frontmatter ---
      const isFrontmatter = Boolean(ast.attributes.frontmatter);
      log.debug(`Frontmatter detected: ${isFrontmatter}`);
      log.debug(`Raw frontmatter content: ${ast.attributes.frontmatter}`);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      const frontmatter: Record<string, unknown> = isFrontmatter
        ? YAML.parse(ast.attributes.frontmatter as string)
        : {};
      log.debug(`Parsed frontmatter: ${JSON.stringify(frontmatter)}`);

      // --- Prepare to load Schemas & Partials ---
      const dependencies: string[] = [];
      let configFromSchema: Config = {};
      let partialsFromSchema: Config["partials"] = {};
      let partialsFromPartials: Config["partials"] = {};

      // --- Discover optional schema directory ---
      const schemaDir = findFirstDirectory(schemaPaths);

      // --- Load Schemas and Partials from resolved schema directory ---
      if (schemaDir) {
        const { config, deps: schemaDeps } = await loadSchemas(schemaDir);
        configFromSchema = config;
        dependencies.push(...schemaDeps);

        const { config: partials, deps } = loadPartials(schemaDir, extensions);
        if (partials) partialsFromSchema = partials;
        dependencies.push(...deps);
      }

      // --- Load Partials from specified partials directory ---
      if (partialsPath) {
        const { config, deps } = loadPartials(partialsPath, extensions, true);
        if (config) partialsFromPartials = config;
        dependencies.push(...deps);
      }

      // --- Assemble full config ---
      const fullConfig: Config = {
        // Start with base config loaded from the schema directory
        // Merge in config passed as options
        nodes: { ...configFromSchema.nodes, ...nodes },
        tags: { ...configFromSchema.tags, ...tags },
        functions: { ...configFromSchema.functions, ...functions },
        // Make $frontmatter available as variable
        variables: { ...configFromSchema.variables, ...variables, frontmatter },
        // Merge partials: explicitly set partials overwrrite auto-loaded ones
        partials: { ...partialsFromSchema, ...partialsFromPartials },
      };

      // --- Validate Markdoc AST ---
      const errors = Markdoc.validate(ast, fullConfig);
      handleValidationErrors(errors, validationLevel, filename);

      // --- Tranform AST with loaded config ---
      const transformedContent = Markdoc.transform(ast, fullConfig);

      // --- Render Markdoc AST to Svelte ---
      const svelteContent = render(transformedContent);

      // --- Define frontmatter string for Svelte ---
      // Extract filename without path and extension
      const baseFilename = filename ? basename(filename, extname(filename)) : '';
            // Always declare module context, include filename and optionally frontmatter
      const scriptModuleTag = 
        `<script module>\n` +
        `\texport const slug = "${baseFilename}";\n` +
        (isFrontmatter 
          ? `\texport const frontmatter = ${JSON.stringify(frontmatter)};\n` +
            `\tconst { ${Object.keys(frontmatter).join(", ")} } = frontmatter;\n`
          : '') +
        `</script>\n`;

      // --- Generate component import statements ---
      const usedSvelteComponentNames = extractUsedSvelteComponents(transformedContent);
      const componentImportStatements = getComponentImports(usedSvelteComponentNames, componentsPath);

      // --- Construct script tag content ---
      let allScriptImports = componentImportStatements;
      if (layoutPath) {
        if (allScriptImports && allScriptImports.trim() !== "") {
          allScriptImports += "\n";
        }
        allScriptImports += `\timport Layout_MARKDOC from '${layoutPath}';\n`;
      }

      const scriptTag = allScriptImports ? `<script>\n${allScriptImports}</script>\n` : "";

      // --- Define layout wrapper strings ---
      // Pass frontmatter to the layout component if it exists
      const layoutWrapperOpen = layoutPath
        ? `<Layout_MARKDOC${isFrontmatter ? ` {...frontmatter}` : ""}>\n`
        : "";
      const layoutWrapperClose = layoutPath ? `\n</Layout_MARKDOC>` : "";

      // --- Assemble final Svelte code ---
      const code =
        scriptModuleTag +
        scriptTag +
        layoutWrapperOpen +
        svelteContent +
        layoutWrapperClose;
      
      return {
        code: code,
        dependencies,
      };
    },
  };
};
