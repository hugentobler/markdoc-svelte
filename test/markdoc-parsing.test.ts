import { markdocPreprocess } from "../src/main.ts";
import { commonMark, markdocBuiltInTags } from "./constants.ts";
import { describe, it, expect } from "vitest";
import type {Options} from "../src/types.ts";

describe("Markdown parsing", () => {
  const testOptions: Options = {
    validationLevel: "warning"
  };
  
  it("CommonMark Tags", async () => {
    expect(
      await markdocPreprocess(testOptions).markup!({ content: commonMark, filename: "test.md" }),
    ).toMatchSnapshot();
  });

  it("Markdoc Table", async () => {
    expect(
      await markdocPreprocess().markup!({
        content: markdocBuiltInTags,
        filename: "test.md",
      }),
    ).toMatchSnapshot();
  });
});
