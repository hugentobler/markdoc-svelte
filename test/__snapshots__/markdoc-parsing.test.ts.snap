// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Markdown parsing > CommonMark Tags 1`] = `
{
  "code": "<script module>
	export const slug = "test";
</script>
<article><h1>Heading Level 1</h1><h2>Heading Level 2</h2><h3>Heading Level 3</h3><h4>Heading Level 4</h4><h5>Heading Level 5</h5><h6>Heading Level 6</h6><h2>Stext Heading</h2><p>This is a heading specifically for testing <strong>strong text</strong> (stext) within a heading.</p><p>This is a paragraph. It contains some text content to demonstrate how paragraphs work in Markdown.</p><hr /><p>This is a horizontal rule (HR) using three dashes.</p><hr /><p>This is a horizontal rule (HR) using three asterisks.</p><hr /><p>This is a horizontal rule (HR) using three underscores.</p><p><img src="https://example.com/image.jpg" alt="Alt text for the image" title="Image Title" /></p><pre data-language="javascript"><code>// Fenced code block with Markdoc brackets &lcub;&rcub; and multiple linesconst obj = &lcub; key: 'value' &rcub;;console.log(\`Hello, &lcub;world&rcub;! Template: $&lcub;obj.key&rcub;\`);function test() &lcub;  return &lcub; nested: true &rcub;;&rcub;</code></pre><blockquote><p>This is a blockquote. It can span multiple lines.</p></blockquote><ol><li>Ordered list item 1</li><li>Ordered list item 2</li></ol><ul><li>Unordered sublist item 1</li><li>Unordered sublist item 2</li></ul><ol start="3"><li>Ordered list item 3</li></ol><ul><li>Unordered list item 1</li><li>Unordered list item 2<ol><li>Ordered sublist item 1</li><li>Ordered sublist item 2</li></ol></li><li>Unordered list item 3</li></ul><p>This is <strong>strong text</strong>. This is <em>emphasized text</em>. This is <s>strikethrough text</s>.</p><p>This is a <a href="https://example.com" title="Link Title">link</a> to an external website. This is a <a href="https://example.com/tricky" title="Title with &lt; &gt; &amp; &quot; ' characters">link with tricky chars in title</a> testing attribute escaping.</p><p>Here is some <code>inline code</code> within a sentence. Here is inline code with brackets: <code>&lcub; example: true &rcub;</code>.</p><p>This is a line of text. This is another line of text after a hard break.</p><p>This is a line of text with a soft break and it continues on the next line.</p><h3>Tabs</h3><p>Here is a line with a tab: This line starts with a tab character.</p><h3>Special Characters</h3><p>Here are some special characters: <code>! @ # $ % ^ &amp; * ( ) - _ + = &lcub; &rcub; [ ] |  : ; &quot; ' &lt; &gt; , . ? / ~</code></p><h3>Backslash-Escaped Characters</h3><p>Here are some backslash-escaped characters: \\ Backslash \` Backtick</p><ul><li>Asterisk _ Underscore { Curly brace open } Curly brace close [ Square bracket open ] Square bracket close ( Parenthesis open ) Parenthesis close</li></ul><h1>Hash</h1><ul><li>Plus</li></ul><ul><li>Minus . Period ! Exclamation mark</li></ul><h3>Valid HTML Entities or Numeric Characters</h3><p>Here are some valid HTML entities and numeric characters: © (©) ® (®) © (© using numeric code) ® (® using numeric code) & (&) < (<) > (>) " (") ' (')</p><h3>Indented Code Block</h3><p>// This is an indented code block with Markdoc brackets {} function greet() { const config = { enabled: true }; console.log("Hello from {indented} code block!", config); }</p><h3>HTML Block</h3><p><div> <p>This is a block of raw HTML.</p> <p>It will render as HTML if supported.</p> </div></p><h3>Indented Code Fence</h3><pre><code>// This is an indented code fence with &lcub;brackets&rcub;console.log(&quot;Indented &lcub;code&rcub; fence example&quot;);</code></pre><h3>Inline Code Example in Another Block</h3><p>Here is an example of inline code within a blockquote:</p><blockquote><p>This is a blockquote with <code>inline code</code> and <code>&lcub; bracketed: code &rcub;</code> inside it.</p></blockquote></article>",
  "dependencies": [],
}
`;

exports[`Markdown parsing > Markdoc Table 1`] = `
{
  "code": "<script module>
	export const slug = "test";
</script>
<article><h3>Basic table</h3><table><thead><tr><th>Heading 1</th><th>Heading 2</th></tr></thead><tbody><tr><td>Row 1 Cell 1</td><td>Row 1 Cell 2</td></tr><tr><td>Row 2 Cell 1</td><td>Row 2 cell 2</td></tr></tbody></table><h3>Table with rich content</h3><table><thead><tr><th>Foo</th><th>Bar</th><th>Baz</th></tr></thead><tbody><tr><td><pre><code>puts &quot;Some code here.&quot;</code></pre></td><td></td><td>Text in a table</td></tr><tr><td><p>A "loose" list with</p><p>multiple line items</p></td><td><p>Test 2</p></td><td><p>Test 3</p></td></tr><tr><td>Test 1</td><td colSpan="2">A cell that spans two columns </td></tr></tbody></table><h3>Table without headings</h3><table><thead></thead><tbody><tr><td>foo</td><td>bar</td></tr><tr><td>foo</td><td>bar</td></tr></tbody></table><h3>Table with col and row span</h3><table><thead></thead><tbody><tr><td>foo</td><td>bar</td></tr><tr><td colSpan="2">foo </td></tr></tbody></table><h3>Table with text alignment</h3><table><thead><tr><th align="center">Column 1 </th><th>Column 2</th><th align="right">Column 3 </th></tr></thead><tbody><tr><td>foo</td><td>bar</td><td>baz</td></tr><tr><td>foo</td><td align="right">bar </td><td>baz</td></tr><tr><td align="center">foo </td><td>bar</td><td>baz</td></tr></tbody></table></article>",
  "dependencies": [],
}
`;
